### PR TITLE
client: Unsubscribe from a topic

### PR DESCRIPTION
**Pull request by besedad**
Adds `Unsubscribe` and `UnsubscribePredefinedTopic` methods to the client.

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -393,12 +393,44 @@ func (c *Client) Subscribe(topic string, qos uint8, callback MessageHandlerFunc)
 	}
 }
 
+func (c *Client) unsubscribe(topicName string, topicIDType uint8, topicID uint16) error {
```
Comment (mprymek) [inline client/client.go:396]: You've placed `unsubscribe()` and `Unsubscribe()` between `Subscribe()` and `SubscribePredefinedTopic()`. I think we should keep all subscribe-related functions together and all unsubscribe-related functions together.

File: client/client_test.go [hunk 2]
```diff
@@ -1354,6 +1354,352 @@ func TestSubscribePredefined(t *testing.T) {
...
+		suback := msgs.NewSubackMessage(1, 0, msgs.RC_ACCEPTED)
+		suback.CopyMessageID(subscribe)
+		stp.send(suback)
+
+		// client <--PUBLISH-- GW
+		msgID := uint16(123)
+		publish := msgs.NewPublishMessage(suback.TopicID,
+			msgs.TIT_REGISTERED, []byte(""), qos, false, false)
+		publish.SetMessageID(msgID)
+		stp.send(publish)
```
Comment (mprymek) [inline client/client_test.go:1390]: This block is not needed IMHO. I think, it's enough to simulate subscribe and then unsubscribe.

If you agree, you can also remove everything related to `callbackFired` below...

File: client/client_test.go [hunk 3]
```diff
@@ -1354,6 +1354,352 @@ func TestSubscribePredefined(t *testing.T) {
...
+	if err := stp.client.Disconnect(); err != nil {
+		stp.t.Fatal(err)
+	}
+	assert.Equal(util.StateDisconnected, stp.client.state.Get())
+	stp.assertClientDone()
+
+	wg.Wait()
+}
+
+func TestUnsubscribeWildcard(t *testing.T) {
```
Comment (mprymek) [inline client/client_test.go:1439]: This test doesn't make sense to me. The most of it tests `PUBLISH` handling.

We can discuss it in detail elsewhere.

File: client/client_test.go [hunk 4]
```diff
@@ -1354,6 +1354,352 @@ func TestSubscribePredefined(t *testing.T) {
...
+
+	select {
+	case <-callbackFired:
+		// ok
+	case <-time.After(time.Second):
+		stp.t.Fatal("subscribe callback not fired")
+	}
+}
+
+func TestUnsubscribeShort(t *testing.T) {
```
Comment (mprymek) [inline client/client_test.go:1541]: As with `TestUnsubscribe`, I would remove everything `callbackFired`-related.

File: client/client_test.go [hunk 5]
```diff
@@ -1354,6 +1354,352 @@ func TestSubscribePredefined(t *testing.T) {
...
+	if err := stp.client.Disconnect(); err != nil {
+		stp.t.Fatal(err)
+	}
+	assert.Equal(util.StateDisconnected, stp.client.state.Get())
+	stp.assertClientDone()
+
+	wg.Wait()
+}
+
+func TestUnsubscribePredefined(t *testing.T) {
```
Comment (mprymek) [inline client/client_test.go:1622]: As with `TestUnsubscribe`, I would remove everything `callbackFired`-related.

File: client/unsubscribe_transaction.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+	switch unsubscribe.TopicIDType {
+	case msgs.TIT_STRING:
+		topicName = string(unsubscribe.TopicName)
+
+		t.log.Debug(`Topic "%s" unsubscribed`,
+			topicName,
+		)
+		t.client.registeredTopicsLock.Lock()
+		delete(t.client.registeredTopics, topicName)
+		t.client.registeredTopicsLock.Unlock()
```
Comment (mprymek) [inline client/unsubscribe_transaction.go:50]: I don't understand this. Why to unregister topic on unsubscribe?

**Review by mprymek — CHANGES_REQUESTED** [review]
I have multiple fundamental comments. I think we should discuss it personally on Monday perhaps.

File: client/client_test.go (hunk 2 above)
Comment (besedad) [reply client/client_test.go]: Done.

File: client/client_test.go (hunk 4 above)
Comment (besedad) [reply client/client_test.go:1420]: Done.

File: client/client_test.go (hunk 5 above)
Comment (besedad) [reply client/client_test.go:1484]: Done.

File: client/unsubscribe_transaction.go (hunk 6 above)
Comment (besedad) [reply client/unsubscribe_transaction.go]: For the record: we have reached an agreement that deleting a registered topic is not correct behavior for the fact that MQTT-SN doesn't mention this. I have deleted the lines in question.

File: client/unsubscribe_transaction.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	var topicName string
+	unsubscribe := t.Data.(*msgs.UnsubscribeMessage)
+
+	switch unsubscribe.TopicIDType {
+	case msgs.TIT_STRING:
+		topicName = string(unsubscribe.TopicName)
+
+		t.log.Debug(`Topic "%s" unsubscribed`,
+			topicName,
+		)
```
Comment (mprymek) [inline client/unsubscribe_transaction.go:47]: This debug log is fired on `TIT_STRING` unsubscribe only. I would move it under the whole `switch` block to be effective for all types of topics.

File: client/client.go [hunk 8]
```diff
@@ -399,6 +399,38 @@ func (c *Client) SubscribePredefinedTopic(topicID uint16, qos uint8, callback Me
...
+// it's treated as a short topic.
+func (c *Client) Unsubscribe(topic string) error {
+	if msgs.IsShortTopic(topic) {
+		return c.unsubscribe("", msgs.TIT_SHORT, msgs.EncodeShortTopic(topic))
+	} else {
+		return c.unsubscribe(topic, msgs.TIT_STRING, 0)
+	}
+}
+
+func (c *Client) UnsubscribePredefinedTopic(topicID uint16) error {
```
Comment (mprymek) [inline client/client.go:430]: IIRC, you proposed to use `UnsubscribePredefined` only. I'd use it.

**Review by mprymek — CHANGES_REQUESTED** [review]
Just two minor bugs left, LGTM otherwise.

File: client/client.go (hunk 8 above)
Comment (besedad) [reply client/client.go:430]: I think it is a separate issue which deserves its own PR. The current naming is compatible with `SubscribePredefinedTopic`.

File: client/unsubscribe_transaction.go (hunk 7 above)
Comment (besedad) [reply client/unsubscribe_transaction.go]: Thanks. Fixed.

File: client/client.go (hunk 8 above)
Comment (mprymek) [reply client/client.go:430]: ok

**Review by mprymek — APPROVED** [review]

**Pull request merged**